docs/extend/plugins_authentication.md

```diff
@@ -0,0 +1,112 @@
+<!--[metadata]>
++++
+title = "Authentication plugins"
+description = "How to authenticate clients with external authentication plugins"
+keywords = ["Examples, Usage, authn, docker, plugin, api"]
+[menu.main]
+parent = "mn_extend"
++++
+<![end-metadata]-->
+
+# Write an authentication plugin
+
+Docker authentication plugins enable Docker daemons to authenticate clients
+using data included in the client HTTP request.  See the [plugin
+documentation](plugins.md) for information about implementing plugins that is
+not specific to authentication plugins.
+
+# Command-line changes
+
+An authentication plugin is enabled by using the `--authn` flag and the
+`--authn-opt plugins` option with the `dockerd` command.
+
+    $ dockerd --authn --authn-opt plugins=htpasswd
+
+# Authentication protocol
+
+If a plugin registers itself as an `Authentication` plugin when activated, then
+it is expected to provide `Authentication.SetOptions` and
+`Authentication.Authenticate` handlers.
+
+### /Authentication.SetOptions
+
+**Request**:
+```
+{
+    "Options": {}
+}
+```
+#### The `Options` map contains the authentication options which were passed
+to the daemon at startup-time.
+
+**Response**:
+```
+{
+}
+```
+
+#### Implementation
+
+Authentication options which were provided to the daemon will be passed to the
+plugin using this endpoint.  Plugins which implement the "Basic" authentication
+scheme should take note of the "realm" option, if one is set.
+
+### /Authentication.Authenticate
+
+**Request**:
+```
+{
+    "Method": "GET"/"POST",
+    "URL": {},
+    "Host": "",
+    "Header": {},
+    "Certificate": ""
+}
+```
+#### The `Method` string is taken from the client request.
+#### The `URL` map contains `Scheme`, `Path`, `Fragment`, `RawQuery`, and
```

Comment:
IMO this should just be a string. There's no need to complicate it by sending a parsed version of the URL.